lib2/client/okta.go

```diff
@@ -0,0 +1,667 @@
+// several different MFA types including:
+//
+// SMS: Okta will send an SMS to the user that includes a code that needs to be
+//      sent back as the verify step.
+// PUSH: Either OKTA verify or DUO are supported.
+// U2F: a u2f hardware token, eg. Yubikey
+//
+// TODO: document full list of MFA supported and create tests
+//
+// More details about the auth flow implemented by this client can be found in
+// Okta documentation: https://developer.okta.com/docs/reference/api/authn
+//
+func (o *OktaClient) AuthenticateUser() (err error) {
+	var payload []byte
+	//TODO(switj): cleanup this marshal struct
+	payload, err = json.Marshal(OktaUser{Username: o.creds.Username, Password: o.creds.Password})
+	if err != nil {
+		return
+	}
+
+	log.Debug("Posting first call to authenticate the user.")
+	res, err := o.Request("POST", "api/v1/authn", url.Values{}, payload, "json", true)
+	if err != nil {
+		return fmt.Errorf("Failed to authenticate with okta. If your credentials have changed, use 'aws-okta add': %#v", err)
+	}
+	defer res.Body.Close()
+
+	err = json.NewDecoder(res.Body).Decode(&o.UserAuth)
+	if err != nil {
+		return
+	}
+	// Step 2 : Challenge MFA if needed
+	if o.UserAuth.Status == "MFA_REQUIRED" {
+		log.Info("Requesting MFA. Please complete two-factor authentication with your second device")
+		if err = o.challengeMFA(); err != nil {
+			return
+		}
+	} else if o.UserAuth.Status == "PASSWORD_EXPIRED" {
+		return fmt.Errorf("Password is expired, login to Okta console to change")
+	}
+
+	if o.UserAuth.SessionToken == "" {
+		log.Debug("Auth failed. Reason: Session token isn't present.")
+		return fmt.Errorf("authentication failed for %s", o.creds.Username)
+	}
+	return
+}
+
+// Public interface to get the Okta session token.
+func (o *OktaClient) GetSessionToken() string {
+	return o.UserAuth.SessionToken
+}
+
+// Validates the provided MFA config matches what the user has configured in
+// Okta. If the provided config doesn't match an error will be returned.
+func selectMFADeviceFromConfig(o *OktaClient) (*OktaUserAuthnFactor, error) {
+	log.Debugf("MFAConfig: %v\n", o.MFAConfig)
+	if o.MFAConfig.Provider == "" || o.MFAConfig.FactorType == "" {
+		return nil, nil
+	}
+
+	for _, f := range o.UserAuth.Embedded.Factors {
+		log.Debugf("%v\n", f)
+		if strings.EqualFold(f.Provider, o.MFAConfig.Provider) && strings.EqualFold(f.FactorType, o.MFAConfig.FactorType) {
+			log.Debugf("Using matching factor \"%v %v\" from config\n", f.Provider, f.FactorType)
+			return &f, nil
+		}
+	}
+
+	return nil, fmt.Errorf("Failed to select MFA device with Provider = \"%s\", FactorType = \"%s\"", o.MFAConfig.Provider, o.MFAConfig.FactorType)
+}
+
+// Will prompt the user to select one of the configured MFA devices if an MFA
+// configuration isn't provided.
+//
+// TODO: convert this to a passed io reader to facilitate simple testing.
+func (o *OktaClient) selectMFADevice() (*OktaUserAuthnFactor, error) {
```

Comment:
I'd abstract this into some sort of `Chooser` interface and have the CLI code pass it in. Separates the UI from the business logic